Adding taptree_of_horror example.

miketwenty1 opened the pull request:
Adding taptree_of_horror example.
~~Adding bitcoin_hashes as dev dependency.~~
~~Made some formatting consistency corrections in Cargo.toml.~~

miketwenty1 commented:
@apoelstra @sanket1729 

Rebased the last commit into my PR. Looking for review/feedback. I laid out a few talking points for discussion so I can get an idea of how to get this merged in:

1. Does the setup of the example work for you?
2. Are you okay with my linter’s consistency edits to `Cargo.toml`?
3. Ok that I'm bringing in `bitcoin_hashes` as a dev-dep? 
4. Would you prefer a themed example referencing TABConf 6's "Taptree of Horror," or a non-themed one? I was thinking themed, and eventually adding the review Youtube video and the Sats Conf video when they are ready https://agenda.satsconf.com.br/satsconf2024/talk/GQC8YK/ , if you would prefer non-themed, let me know what you're thinking for a example and folder name.
5. I included the `.excalidraw` file in case the tree graphic needed to be edited, but I can remove it if not needed.

apoelstra commented:
> * Are you okay with my linter’s consistency edits to `Cargo.toml`?

I'd rather not. It seems like you just need to add some lines to the bottom for the new example and that's it. (Also ironically, your new source files have some trailing spaces in them, so maybe these need to have a linter/formatter run on them.)

> * Ok that I'm bringing in `bitcoin_hashes` as a dev-dep?

No, this doesn't seem necessary to me. I tried removing this and replacing `bitcoin_hashes::Hash` with `bitcoin::hashes::Hash` in your code and everything seemed to compile fine.

> * Would you prefer a themed example referencing TABConf 6's "Taptree of Horror," or a non-themed one? I was thinking themed

Yeah, I like the theme.

miketwenty1 commented:
> I'd rather not. 

Fixed

> (Also ironically, your new source files have some trailing spaces in them, so maybe these need to have a linter/formatter run on them.)

I'm not sure what you mean. Can you give me a specific example.


> No, this doesn't seem necessary to me. I tried removing this and replacing `bitcoin_hashes::Hash` with `bitcoin::hashes::Hash` in your code and everything seemed to compile fine.

Fixed

Thanks for the feedback, fixed the issues you pointed. Let me know what else you'd like to see in this PR.

apoelstra commented on examples/taptree_of_horror/README.md:
```diff
@@ -0,0 +1,20 @@
+### Running this example:
+- `cargo run --example taptree_of_horror --features "compiler"`
+
+### Originally based on the TABConf 6, CTB. 
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

trailing space here

apoelstra commented on examples/taptree_of_horror/helper_fns.rs:
```diff
@@ -0,0 +1,57 @@
+}
+
+pub fn produce_kelly_hash(secret: &str) -> (sha256::Hash, sha256::Hash) {
+    let prepreimage: Vec<u8> = secret.as_bytes().to_vec();
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

No need to call `to_vec` here.

apoelstra commented on examples/taptree_of_horror/helper_fns.rs:
```diff
@@ -0,0 +1,57 @@
+        let derived_xpriv: Xpriv = xprv.xkey.derive_priv(secp, &derivation_path).unwrap();
+
+        pks.insert(i.try_into().unwrap(), pk);
+        prvs.insert(i.try_into().unwrap(), derived_xpriv);
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

Both of these `insert` calls can be replaced with `push`. e.g. `pks.push(pk); prvs.push(derived_xpriv)`.

apoelstra commented on examples/taptree_of_horror/helper_fns.rs:
```diff
@@ -0,0 +1,57 @@
+        _ => panic!("not an xpriv"),
+    };
+
+    for i in 0..9 {
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

I would suggest lifting this to a constant `KEYS_PER_CHARACTER` or something.

miketwenty1 replied:
Went with `KEYS_PER_PERSONA`

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+    let internal_xpub: miniscript::DescriptorPublicKey =
+        internal_desc_secret.to_public(secp).unwrap();
+    let _intneral_xpriv: DescriptorXKey<Xpriv> = match internal_desc_secret.clone() {
+        DescriptorSecretKey::XPrv(x) => Some(x),
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

I would change the spelling of `intneral_xpriv`, move the `.clone` into the `Some`, and comment this whole statement out (since the place that it's used is also commented out).

In general I think it's fine to have commented-out-code in examples. cc @sanket1729 what do you think?

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+
+    // define individual variables to make it easy to insert into the policy string.
+    // Define individual variables with pattern matching in a compact format
+    let (a0, a1, a2, a3, a4, a5, a6, a7, a8) = match &a_pks[..] {
```
In  1282d30653a3cb4e188e59d5a3dd8de909aa9477

Can replace all these with one-liners, e.g.

```rust
let [a0, a1, a2, a3, a4, a5, a6, a7, a8] = a_pks[..].try_into().unwrap();
```

I would also drop the `Define individual variables with pattern matching` comment since it's just a repetition of the code.

miketwenty1 replied:
fyi, I'm also be including the explicit type to make the compiler/linter happy
`[PublicKey; KEYS_PER_PERSONA]`

apoelstra replied:
If you think that's necessary -- though I tried it without type annotations and it seemed to work. (Maybe you need the annotations once you convert all of them; I only did one.)

miketwenty1 replied:
I tried it again and removed the type for the "alice"/"a" keys and got this:

```
cargo +1.63.0 run --example taptree_of_horror --features compiler
   Compiling miniscript v12.2.0 (/Users/tzez/git/github/rust-bitcoin/rust-miniscript)
error[E0282]: type annotations needed
  --> examples/taptree_of_horror/taptree_of_horror.rs:88:9
   |
88 |     let [a0, a1, a2, a3, a4, a5, a6, a7, a8] = a_pks[..].try_into().unwrap();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: type must be known at this point
help: consider giving this pattern a type
   |
88 |     let [a0, a1, a2, a3, a4, a5, a6, a7, a8]: _ = a_pks[..].try_into().unwrap();
   |                                             +++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `miniscript` due to previous error
```

having the specific type solves this error, if you want to address it in a different way let me know.

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+                and(
+                    thresh(10, pk({a0}), pk({b0}), pk({c0}), pk({d0}), pk({e0}), pk({f0}), pk({h0}), pk({i0}), pk({j0}), pk({l0})),
+                    thresh(3, sha256({k}), ripemd160({g}), after({oct_23_morning}))
+                ),                               
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

trailing whitespace here

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+    let derived_descriptor = policy_desc.at_derivation_index(0).unwrap();
+    let _script_address = derived_descriptor.address(Network::Regtest).unwrap();
+    // println!("the receiving address of this script is: {}", _script_address);
+    // println!("\ndescriptor is: {}\n", desc);
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

I think you should uncomment these. It's nice to see some output from examples.

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+
+    let prev_amount = Amount::from_sat(100_000_000);
+    let witness_utxo = TxOut {
+        value: prev_amount, // 10 BTC, has to be exactly the same amount as what's in the txid being used?
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

Yep, it does.

miketwenty1 replied:
left over comments from a younger me. removing.

apoelstra commented on examples/taptree_of_horror/taptree_of_horror.rs:
```diff
@@ -0,0 +1,314 @@
+    let oct_23_morning = "1729692000"; // Oct 23, 10:00 AM EST
+    let oct_24_evening = "1729819800"; // Oct 24, 09:30 PM EST
+    let oct_25_afternoon = "1729877400"; // Oct 25, 01:30 PM EST
+    let oct_26_morning = "1729942200"; // Oct 26, 07:30 AM EST
```
In 1282d30653a3cb4e188e59d5a3dd8de909aa9477:

All these can just be numbers rather than strings. (It'll save you some .parse calls below).

apoelstra commented:
Super cool! Done reviewingc1282d30653a3cb4e188e59d5a3dd8de909aa9477. Lots of comments but they are all super minor things.

apoelstra commented:
89cdeca33d1e1f93a27c396f9a5ec475e28ff8a8 looks great to me! CI is showing a clippy lint you can address.

apoelstra commented:
utack 08cea1cdd3e58860ba35a2082fe0759c11605669

apoelstra commented:
utack 6910eb4b76bb84e53aea65a7bb707e27516a473f

apoelstra reviewed (approved):
ACK 6910eb4b76bb84e53aea65a7bb707e27516a473f; successfully ran local tests

apoelstra commented:
cc @sanket1729 do you want to review this? I'll hold off on merging til tomorrow afternoon but I think this is good to go in.

sanket1729 reviewed (approved):
ACK 6910eb4b76bb84e53aea65a7bb707e27516a473f

sanket1729 reviewed (approved):
Did not a careful code review. The psbt APIs are used correctly and will serve as a good reference point for future users. 

Maybe, we have a good first issue to simplify this example and create a simple psbt sign/update/finalize flow for a simpler policy :) .

apoelstra commented:
We need to update `contrib/test_vars.sh` to include this example. (But really, we need to update CI to just read this stuff out of the toml file.)
